lib/fluent/plugin/in_tail.rb

```diff
@@ -386,78 +386,75 @@ def setup_watcher(path_with_inode, pe)
             begin
-              pe.update(Fluent::FileWrapper.stat(path).ino, 0)
+              pe.update(Fluent::FileWrapper.stat(target_info.path).ino, 0)
             rescue Errno::ENOENT
               $log.warn "#{path} not found. Continuing without tailing it."
             end
           end
         end
 
         begin
-          tw = setup_watcher(path_with_inode, pe)
+          tw = setup_watcher(target_info, pe)
         rescue WatcherSetupError => e
           log.warn "Skip #{path} because unexpected setup error happens: #{e}"
           next
         end
-        target_info = TargetInfo.new(path, Fluent::FileWrapper.stat(path).ino)
+        target_info = TargetInfo.new(target_info.path, Fluent::FileWrapper.stat(target_info.path).ino)
         @tails[target_info] = tw
       }
     end
 
-    def stop_watchers(paths_with_inodes, immediate: false, unwatched: false, remove_watcher: true)
-      paths_with_inodes.each_value { |path_with_inode|
+    def stop_watchers(targets_info, immediate: false, unwatched: false, remove_watcher: true)
+      targets_info.each_value { |target_info|
         if remove_watcher
-          tw = @tails.delete(path_with_inode)
+          tw = @tails.delete(target_info)
         else
-          tw = @tails[path_with_inode]
+          tw = @tails[target_info]
         end
         if tw
           tw.unwatched = unwatched
           if immediate
-            detach_watcher(tw, path_with_inode.ino, false)
+            detach_watcher(tw, target_info.ino, false)
           else
-            detach_watcher_after_rotate_wait(tw, path_with_inode.ino)
+            detach_watcher_after_rotate_wait(tw, target_info.ino)
           end
         end
       }
     end
 
     def close_watcher_handles
-      @tails.keys.each do |path_with_inode|
-        tw = @tails.delete(path_with_inode)
+      @tails.keys.each do |target_info|
+        tw = @tails.delete(target_info)
         if tw
           tw.close
         end
       end
     end
 
     # refresh_watchers calls @tails.keys so we don't use stop_watcher -> start_watcher sequence for safety.
-    def update_watcher(path_with_inode, pe)
+    def update_watcher(target_info, pe)
       log.info("detected rotation of #{path}; waiting #{@rotate_wait} seconds")
 
       if @pf
-        unless pe.read_inode == @pf[path_with_inode.path, pe.read_inode].read_inode
+        target_info_from_position_entry = TargetInfo.new(target_info.path, pe.read_inode)
+        unless pe.read_inode == @pf[target_info_from_position_entry].read_inode
           log.debug "Skip update_watcher because watcher has been already updated by other inotify event"
           return
         end
       end
 
-      target_info = TargetInfo.new(path_with_inode.path, pe.read_inode)
       rotated_tw = @tails[target_info]
-
-      new_target_info = TargetInfo.new(path_with_inode.path, path_with_inode.ino)
+      new_target_info = TargetInfo.new(target_info.path, target_info.ino)
```

Comment:
`target_info.dup` has a problem?

Reply:
No issue.
Fixed in https://github.com/fluent/fluentd/pull/3196/commits/828f91496763e8069a6e3a1c082a1f8d0b492d04.